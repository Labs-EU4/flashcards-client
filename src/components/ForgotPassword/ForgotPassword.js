import React, { useState } from 'react';
// import '../../App.css'
// import { Form, Icon, Input, Button } from 'antd';

const ForgotPassword = () => {
    
    const [formValues, setFormValues] = useState({
        email: ""
    })

    const handleChange = e => {
        setFormValues({
            formValues,
            [e.target.name]: e.target.value
          })
    }
    
    const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
    let errorcount = 0

    const handleSubmit = e => {
        e.preventDefault();
        console.log(formValues.email);
        if(emailRegex.test(formValues.email) === true) {
            console.log("good")
            setFormValues({
                    email: ""
                })
        } else {
            console.log("BAD");
            const p = document.createElement("p")
            if(errorcount < 1) {
                errorcount++
                p.textContent = "Bad Email Address"
                p.style.color = "red"
                document.querySelector("body").append(p)
            }
        }
    }

  return (
    <div className="forgot-password" onSubmit={handleSubmit}>
        <h1>Reset Password</h1>

        <form>
            <label>Email</label>
            <input type="text" name="email" value={formValues.email} onChange={handleChange} />
            <input type="submit" />
        </form>
        {/* <Form>
            <Input>
            </Input>
        </Form> */}
    </div>
  )
}

export default ForgotPassword;