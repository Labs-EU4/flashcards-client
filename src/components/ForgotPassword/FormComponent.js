import React, {useState} from "react";
import {Form, Icon, Input, Button} from "antd";
import axios from "axios";

const NormalLoginForm = props => {
  //State for the form values
  const [formValues, setFormValues] = useState({
    email: "",
  });

  const handleSubmit = e => {
    e.preventDefault();
    axios
      .post("http://localhost:4003/api/auth/forgot_password", {
        email: formValues.email,
      })
      .then(function(response) {
        setFormValues({
          email: "",
        });
        console.log(response);
      })
      .catch(function(error) {
        console.log(error);
      });
  };

  //necessary to change state values
  const handleChange = e => {
    setFormValues({
      ...formValues,
      [e.target.name]: e.target.value,
    });
  };

  const {getFieldDecorator} = props.form;
  return (
    <Form onSubmit={handleSubmit} className="login-form">
      <Form.Item>
        {getFieldDecorator("email", {
          //rules are for the form validation
          rules: [
            {required: true, message: "Please input a valid email!"},
            {
              type: "email",
              message: "Invalid email",
            },
          ],
        })(
          <Input
            name="email"
            setFieldsValue={formValues.email}
            onChange={handleChange}
            //form icon in the email field, change type for different icons, see antdesign docs
            prefix={<Icon type="mail" style={{color: "rgba(0,0,0,.25)"}} />}
            placeholder="Email"
          />
        )}
      </Form.Item>
      <Form.Item>
        <Button type="primary" htmlType="submit" className="login-form-button">
          Reset
        </Button>
      </Form.Item>
    </Form>
  );
};
//necessary for ant design functionality, reasoning in docs
const WrappedNormalLoginForm = Form.create({name: "normal_login"})(NormalLoginForm);

export default WrappedNormalLoginForm;
