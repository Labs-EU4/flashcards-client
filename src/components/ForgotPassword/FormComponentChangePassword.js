import React, {useState} from "react";
import {useHistory} from "react-router-dom";
import axios from "axios";
import {Form, Input, Button, Icon, Alert} from "antd";

const ResetPasswordForm = props => {
  const history = useHistory();
  const tokenArray = history.location.pathname.split("/");
  const token = tokenArray[tokenArray.length - 1];
  const [state, setState] = useState({
    confirmDirty: false,
    isLoading: false,
    tokenInvalid: null,
  });
  const [formValues, setFormValues] = useState({
    newPassword: "",
  });

  const handleChange = e => {
    setFormValues({
      ...formValues,
      [e.target.name]: e.target.value,
    });
  };

  const handleSubmit = e => {
    e.preventDefault();
    setState({...state, isLoading: true});
    props.form.validateFieldsAndScroll((err, values) => {
      if (!err) {
        axios
          .post(
            `https://flashdecks-staging.herokuapp.com/api/auth/reset_password/${token}`,
            {
              password: formValues.newPassword,
              confirmPassword: formValues.newPassword,
            }
          )
          .then(res => {
            console.log(res);
            setState({...state, isLoading: false, tokenInvalid: false});
            setTimeout(() => {
              history.push("/login");
            }, 2000);
          })
          .catch(err => {
            setState({...state, isLoading: false, tokenInvalid: true});
            console.log(err);
          });
      } else {
        setState({...state, isLoading: false});
      }
    });
  };

  const handleConfirmBlur = e => {
    const {value} = e.target;
    setState({...state, confirmDirty: state.confirmDirty || !!value});
  };

  const compareToFirstPassword = (rule, value, callback) => {
    const {form} = props;
    if (value && value !== form.getFieldValue("password")) {
      callback("The passwords have to match!");
    } else {
      callback();
    }
  };

  const validateToNextPassword = (rule, value, callback) => {
    const {form} = props;
    if (value && state.confirmDirty) {
      form.validateFields(["confirm"], {force: true});
    }
    callback();
  };

  const {getFieldDecorator} = props.form;

  return (
    <div className="form-container">
      <Form onSubmit={handleSubmit} className="new-password-form">
        <h1>Reset Password</h1>
        <Form.Item hasFeedback>
          {getFieldDecorator("password", {
            rules: [
              {
                required: true,
                message: "Password must be at least 5 characters",
                min: 5,
              },
              {
                validator: validateToNextPassword,
              },
            ],
          })(
            <Input.Password
              placeholder="New Password"
              name="newPassword"
              onChange={handleChange}
              prefix={<Icon type="lock" style={{color: "rgba(0,0,0,.25)"}} />}
            />
          )}
        </Form.Item>
        <Form.Item hasFeedback>
          {getFieldDecorator("confirm", {
            rules: [
              {
                required: true,
                message: "Please confirm your password!",
              },
              {
                validator: compareToFirstPassword,
              },
            ],
          })(
            <Input.Password
              placeholder="Confirm Password"
              onBlur={handleConfirmBlur}
              prefix={<Icon type="lock" style={{color: "rgba(0,0,0,.25)"}} />}
            />
          )}
        </Form.Item>
        <Form.Item>
          <Button type="primary" htmlType="submit" loading={state.isLoading}>
            Reset
          </Button>
        </Form.Item>
        {state.tokenInvalid ? (
          <Alert
            message="Token invalid"
            description="The token you tried to use is invalid."
            type="error"
          />
        ) : state.tokenInvalid === false ? (
          <Alert
            message="Success"
            description="Your password was updated!"
            type="success"
          />
        ) : null}
      </Form>
    </div>
  );
};

const WrappedNormalLoginForm = Form.create({name: "normal_login"})(ResetPasswordForm);

export default WrappedNormalLoginForm;
