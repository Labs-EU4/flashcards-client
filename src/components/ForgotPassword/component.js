import React, {useState} from "react";
import "antd/dist/antd.css";
import {Form, Icon, Input, Button} from "antd";

const NormalLoginForm = props => {
  const [formValues, setFormValues] = useState({
    email: "",
  });

  //   const handleSubmit = e => {
  //     e.preventDefault();
  //     props.form.validateFields((err, values) => {
  //       if (!err) {
  //         console.log("Received values of form: ", values);
  //       }
  //     });
  //   };

  const handleChange = e => {
    setFormValues({
      formValues,
      [e.target.name]: e.target.value,
    });
  };
  const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

  const handleSubmit = e => {
    e.preventDefault();
  };

  const {getFieldDecorator} = props.form;
  return (
    <Form onSubmit={handleSubmit} className="login-form">
      <Form.Item
        validateStatus={
          emailRegex.test(formValues.email) === false ? "Error" : "validating"
        }
      >
        {getFieldDecorator("email", {
          rules: [
            {required: true, message: "Please input a valid email!"},
            {
              type: "email",
              message: "Invalid email",
            },
          ],
        })(
          <Input
            name="email"
            setFieldsValue={formValues.email}
            onChange={handleChange}
            prefix={<Icon type="mail" style={{color: "rgba(0,0,0,.25)"}} />}
            placeholder="Email"
          />
        )}
      </Form.Item>
      <Form.Item>
        <Button type="primary" htmlType="submit" className="login-form-button">
          Reset
        </Button>
      </Form.Item>
    </Form>
  );
};

const WrappedNormalLoginForm = Form.create({name: "normal_login"})(NormalLoginForm);

export default WrappedNormalLoginForm;
