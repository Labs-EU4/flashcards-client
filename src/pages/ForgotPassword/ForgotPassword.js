import React from "react";

import NormalLoginForm from "../../components/ForgotPassword/FormComponent";
import "./index.css";

const ForgotPassword = props => {
  return (
    <div className="forgot-password">
      <h1>Reset Password</h1>
      <NormalLoginForm />
    </div>
  );
};

export default ForgotPassword;
