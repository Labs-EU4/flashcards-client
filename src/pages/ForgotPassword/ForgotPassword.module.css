.login-form, .new-password-form {
    max-width: 350px;
    width: 100%;
  }

.ant-form-explain {
      color: red;
  }

.forgotPassword {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 30vh;
  overflow-y: hidden;
}

.forgotPassword h1{
      margin: 0 auto;
      text-align: center;
      margin-bottom: 30px;
      font-size: 2rem;
}

.forgotPassword button {
  width: 100%;
}
