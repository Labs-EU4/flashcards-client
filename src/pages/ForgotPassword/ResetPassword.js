import React from "react";
import FormComponentChange from "../../components/ForgotPassword/FormComponentChangePassword";
import styles from "./ForgotPassword.module.css";

const ResetPassword = props => {
  return (
    <div className={styles.forgotPassword}>
      <FormComponentChange />
    </div>
  );
};

export default ResetPassword;
