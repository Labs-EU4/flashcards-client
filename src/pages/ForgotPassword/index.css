.login-form, .new-password-form {
    max-width: 350px;
    width: 100%;
  }

  .ant-form-explain {
      color: red;
  }

.forgot-password {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 30vh;
}

.forgot-password h1{
      margin: 0 auto;
      text-align: center;
      margin-bottom: 30px;
      font-size: 2rem;
}

.ant-btn-primary {
  width: 100%;
}
