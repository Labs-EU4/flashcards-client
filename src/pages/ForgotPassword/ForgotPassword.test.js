import React from "react";
import * as rtl from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import ForgotPassword from "./ForgotPassword.js";

afterEach(rtl.cleanup);

let wrapper;

beforeEach(() => {
  wrapper = rtl.render(<ForgotPassword />);
});

describe("is the component rendering correctly", () => {
  // it("renders fine without crashing", () => {
  //   expect(wrapper.container).toMatchSnapshot();
  // });

  it("renders the heading", () => {
    let queryValue = wrapper.queryByText(/Reset Password/);
    expect(queryValue).toBeDefined();
  });

  it("renders the form placeholder", () => {
    let queryValue = wrapper.getByPlaceholderText(/Email/);
    expect(queryValue).toBeDefined();
  });

  it("renders the button", () => {
    let button = wrapper.queryByText(/^Reset$/i);
    expect(button).toBeDefined();
  });
});

describe("the input field is working", () => {
  it("responds to text input", () => {
    let input = wrapper.getByPlaceholderText(/Email/);
    rtl.fireEvent.click(input);
    rtl.fireEvent.keyDown(input, {key: "A", code: 65, charCode: 65});
    rtl.fireEvent.keyDown(input, {key: "B", code: 66, charCode: 66});
    rtl.fireEvent.keyDown(input, {key: "C", code: 67, charCode: 67});
    let formValue = wrapper.queryByText(/ABC/);
    expect(formValue).toBeDefined();
  });

  it("shows an validation error if input field is empty", () => {
    let input = wrapper.getByPlaceholderText(/Email/);
    rtl.fireEvent.click(input);
    rtl.fireEvent.keyDown(input, {key: "A", code: 65, charCode: 65});
    rtl.fireEvent.keyDown(input, {key: "B", code: 66, charCode: 66});
    rtl.fireEvent.keyDown(input, {key: "C", code: 67, charCode: 67});
    rtl.fireEvent.keyDown(input, {key: "backspace", code: 8});
    rtl.fireEvent.keyDown(input, {key: "backspace", code: 8});
    rtl.fireEvent.keyDown(input, {key: "backspace", code: 8});
    let error = wrapper.queryByText(/Please input a valid email!/);
    expect(error).toBeDefined();
  });

  it("shows an validation error if input field has an invalid email", () => {
    let input = wrapper.getByPlaceholderText(/Email/);
    rtl.fireEvent.click(input);
    rtl.fireEvent.keyDown(input, {key: "A", code: 65, charCode: 65});
    rtl.fireEvent.keyDown(input, {key: "B", code: 66, charCode: 66});
    rtl.fireEvent.keyDown(input, {key: "C", code: 67, charCode: 67});
    let error = wrapper.queryByText(/Invalid email/);
    expect(error).toBeDefined();
  });
});
