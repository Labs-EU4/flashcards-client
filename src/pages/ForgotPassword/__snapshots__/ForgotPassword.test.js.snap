// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is the component rendering correctly renders fine without crashing 1`] = `
<div>
  <div
    class="forgot-password"
  >
    <h1>
      Reset Password
    </h1>
    <form
      class="ant-form ant-form-horizontal login-form"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <span
                class="ant-input-affix-wrapper"
              >
                <span
                  class="ant-input-prefix"
                >
                  <i
                    aria-label="icon: mail"
                    class="anticon anticon-mail"
                    style="color: rgba(0, 0, 0, 0.25);"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="mail"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0 0 68.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
                      />
                    </svg>
                  </i>
                </span>
                <input
                  class="ant-input"
                  data-__field="[object Object]"
                  data-__meta="[object Object]"
                  id="normal_login_email"
                  name="email"
                  placeholder="Email"
                  setfieldsvalue=""
                  type="text"
                  value=""
                />
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <button
                class="ant-btn login-form-button ant-btn-primary"
                type="submit"
              >
                <span>
                  Reset
                </span>
              </button>
            </span>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
